Adds suggestions for categories and formats to LinkControl

draganescu opened the pull request:
## Description
Closes #20924 and closes #20165 

## How has this been tested?
Tested locally by:

0. Make sure you are:
a) having some categories setup
b) on a theme that supports post formats (twentyfourteen)
c) some posts saved as a post formats other than standard
1. Add a new post
2. Add a navigation block in that post
3. Add a new navigation link
4. Search for something that is included in a category's or a post format's name
5. Observe the category and/or post format is included in suggestions

## Types of changes
- adds a format search handler
- adds a category search handler
- hooks the new handlers in the search endpoint
- updates LinkControl's fetchLinkSuggestions to handle formats and categories as well

**PS**
I don't know if this is the best way to implement this feature. However, because we need the link to the format index page, I can't see a better way to get it than from an API call. We could output formats in the bulk data which is rendered on load, but categories could be too many for this option. 

cc @getdave - this updates indirectly `LinkControl` to show more types of suggestions. Suggestions welcome! :)

draganescu committed:
adds suggestions for categories and formats

adds a format search handler
adds a category search handler
hooks the new handlers in the search endpoint
updates LinkControl's fetchLinkSuggestions to handle formats and categories as well

adamziel commented on packages/editor/src/components/provider/index.js:
```diff
@@ -50,7 +50,25 @@ const fetchLinkSuggestions = async ( search, { perPage = 20 } = {} ) => {
+		} ),
+	} );
+
+	const results = concat( posts, categories, formats );
```
Maybe sort them together? Otherwise categories will always come after posts, and formats after categories.

draganescu replied:
sort by what? we it is "rarer" to link to index pages nowadays :) so they may well be last?

adamziel replied:
I don't have a strong preference here, just bringing it up - I trust your judgement :-)

adamziel commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,104 @@
+		$found_ids = [];
+		foreach ( $format_fake_ids as $format_fake_id => $format_slug ) {
+			$format_title       = $formats[ $format_slug ];
+			$format_slug_match  = stripos( $format_slug, $format_search ) !== false;
+			$format_title_match = stripos( $format_title, $format_search ) !== false;
```
Is there any way to include this in the actual query instead of matching in PHP?

draganescu replied:
there is no query. formats are hardcoded strings

adamziel commented on lib/class-wp-rest-category-search-handler.php:
```diff
@@ -0,0 +1,104 @@
+
+		$found_ids = [];
+		foreach ( $categories as $category ) {
+			$name_match = stripos( $category->name, $category_search ) !== false;
```
Is there any way to include this in the actual query instead of matching in PHP?

draganescu replied:
[Apparently](https://developer.wordpress.org/reference/classes/wp_term_query/__construct/) It is!

adamziel reviewed (approved):
@draganescu I left a few minor notes but nothing big. You'll need to run this through a linter too to make travis happy. What worked for me in the past was `npm run lint-php`. Overall this looks good and works so 👍

talldan commented on packages/editor/src/components/provider/index.js:
```diff
@@ -50,7 +50,25 @@ const fetchLinkSuggestions = async ( search, { perPage = 20 } = {} ) => {
+		} ),
+	} );
+
+	const formats = await apiFetch( {
```
I think the way the `awaits` are specified here will result in the API requests being made serially.

This could be a good opportunity to use `await Promise.all( ... )` to perform the requests in parallel.

draganescu committed:
adds concurent search for suggestions

also searches for category matches via term query and limits formats with no items

TimothyBJacobs commented:
I think implementing this handling as a search handler makes a lot of sense, but I think the categories handler should be generic to any taxonomy, the same way `WP_REST_Post_Search_Handler` does. I think the post formats can stay the way that it is.

We'll also want to implement links the same way the posts handler does but to the individual terms.

spacedmonkey commented:
Unit tests would be good here

What about tags?

rickalee commented:
Should this include all taxonomies and terms with show_in_rest = true as well as Authors/Users?

draganescu commented:
Yes @rickalee I'm working on making the new endpoint generic for all terms :)

talldan commented on packages/edit-navigation/src/index.js:
```diff
@@ -43,12 +43,30 @@ const fetchLinkSuggestions = async ( search, { perPage = 20 } = {} ) => {
-		title: decodeEntities( post.title ) || __( '(no title)' ),
-		type: post.subtype || post.type,
-	} ) );
+	const categories = await apiFetch( {
+		path: addQueryArgs( '/wp/v2/search', {
+			search,
+			per_page: perPage,
+			type: 'category',
+		} ),
+	} );
+
+	const formats = await apiFetch( {
+		path: addQueryArgs( '/wp/v2/search', {
+			search,
+			per_page: perPage,
+			type: 'post-format',
+		} ),
+	} );
```
The PR might be taking a different approach, but I noticed the `await`s need to be removed here to make the `Promise.all` work.

goaround commented:
I would suggest adding a Type chooser to the LinkControl that you can explicitly search for a Post (preselected), Term (Category, Tags or public Custom Post Types) and that developer can register their own Search Handler, and they will be shown in the LinkControl UI. 

You can get all types from `/wp-json/` --> `/wp/v2/search` --> `endpoints` --> `args` --> `type` and even `subtype`

I created a little Plugin as a Gist: https://gist.github.com/goaround/3a91ed216453bb19b9da69a9643468fa which registers a WP_REST_Taxonomy_Search_Handler and intercepts the search request if the query starts with `-ter` for Term e.g. `-ter WordPress`. More information (in german) on my blog: https://go-around.de/blog/interne-link-suche-wordpress-block-editor/

draganescu committed:
generalizes the category search handler to all terms

draganescu committed:
lint

draganescu commented:
I've changed the Category search handler into a Terms search handler that now looks for `search` in all the public taxonomies.

I still need to add tests to this PR.

TimothyBJacobs commented:
We'll also need to add links to `prepare_links` for each item returned. You can take a look at the post handler for an example.

draganescu committed:
adds links to terms search results

draganescu committed:
adds default listing for empty searches and a stup test class

TimothyBJacobs commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+
+		return array(
+			self::RESULT_IDS   => $found_ids,
+			self::RESULT_TOTAL => count( $found_ids ),
```
This is meant to be the total found results, regardless of pagination. So if there were 150 matching terms, this would return `150` even if the user only requested 10.

noisysocks replied:
Done in 07dd970e92.

TimothyBJacobs commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+			$taxonomies = $this->subtypes;
+		}
+
+		$terms = get_terms(
```
We need to implement pagination here.

noisysocks replied:
Done in 07dd970e92.

TimothyBJacobs commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+			$term_search = $request['search'];
+		}
+
+		$term_search = apply_filters( 'rest_term_search_query', $term_search, $request );
```
So this filter should be filtering the list of args passed to `get_terms`, not just the search term.

noisysocks replied:
Done in 07dd970e92.

draganescu committed:
shows empty terms in results for search and tests pass

noisysocks commented:
I used this PR to implement https://github.com/WordPress/gutenberg/pull/24670 and it seems to work pretty well!

noisysocks commented:
Could we rename `post_tag` to `tag`? Or is that what WordPress calls a tag?

noisysocks commented on packages/edit-navigation/src/index.js:
```diff
@@ -44,20 +44,38 @@ function disableInsertingNonNavigationBlocks( settings, name ) {
  * @return {Promise<Object[]>} List of suggestions
  */
 async function fetchLinkSuggestions( search, { perPage = 20 } = {} ) {
```
Can remove `async` now that there's no more `await`s.

noisysocks replied:
Done in 60825881dd.

noisysocks commented on packages/edit-navigation/src/index.js:
```diff
@@ -44,20 +44,38 @@ function disableInsertingNonNavigationBlocks( settings, name ) {
+	} );
+
+	return Promise.all( [ posts, terms, formats ] ).then( ( results ) => {
+		return map( flatten( results ), ( post ) => ( {
```
This is no longer only a `post`, so we should call it something else.

noisysocks replied:
Done in 60825881dd.

noisysocks commented on packages/editor/src/components/provider/index.js:
```diff
@@ -42,20 +42,38 @@ import ConvertToGroupButtons from '../convert-to-group-buttons';
  * @return {Promise<Object[]>} List of suggestions
  */
 const fetchLinkSuggestions = async ( search, { perPage = 20 } = {} ) => {
```
Same here. Can drop `async`.

noisysocks replied:
Done in 60825881dd.

noisysocks commented on packages/editor/src/components/provider/index.js:
```diff
@@ -42,20 +42,38 @@ import ConvertToGroupButtons from '../convert-to-group-buttons';
+	} );
+
+	return Promise.all( [ posts, terms, formats ] ).then( ( results ) => {
+		return map( flatten( results ), ( post ) => ( {
```
Same here. Rename `post`.

noisysocks replied:
Done in 60825881dd.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+/**
+ * REST API: WP_REST_Post_Format_Search_Handler class
+ *
+ * @package WordPress
```
Should be Gutenberg, at least for now.

noisysocks replied:
Done in 5a6edbbcec.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+ *
+ * @package WordPress
+ * @subpackage REST_API
+ * @since 5.5.0
```
Should be 8.9, at least for now.

noisysocks replied:
Removed in 5a6edbbcec.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+			$format_search = $request['search'];
+		}
+
+		$format_search = apply_filters( 'rest_post_format_search_query', $format_search, $request );
```
Needs a doc comment.

noisysocks replied:
Addressed in 1d6e65ae1b.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+	public function search_items( WP_REST_Request $request ) {
+
+		$formats         = get_post_format_strings();
+		$format_fake_ids = array_keys( $formats );
```
Why `fake_ids`? `$format` seems to be what we call it elsewhere in Core.

noisysocks replied:
Addressed in 1d6e65ae1b.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+		$format_search = apply_filters( 'rest_post_format_search_query', $format_search, $request );
+
+		$found_ids = array();
+		foreach ( $format_fake_ids as $format_fake_id => $format_slug ) {
```
We should skip this if `$format_search` is undefined.

noisysocks replied:
Addressed in 1d6e65ae1b.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+	 * @return array Links for the given item.
+	 */
+	public function prepare_item_links( $id ) {
+		return array();
```
Do we have a REST API endpoint for formats? If so should set `self` here.

noisysocks replied:
We do not.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+		$formats          = get_post_format_strings();
+		$format_fake_ids  = array_keys( $formats );
+		$format           = $formats[ $format_fake_ids[ $id ] ];
+		$format_index_url = get_post_format_link( $format_fake_ids[ $id ] );
```
I'd expect this to be named `$format_link` or something.

noisysocks replied:
Fixed in 1d6e65ae1b.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,116 @@
+
+		return array(
+			self::RESULT_IDS   => $found_ids,
+			self::RESULT_TOTAL => count( $found_ids ),
```
This should be the total amount of formats with no filtering.

noisysocks replied:
My mistake: it should include filtering but not pagination. Done in 1d6e65ae1b.

noisysocks commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,163 @@
+	protected function detect_rest_item_route( $term ) {
+		$taxonomy = get_taxonomy( $term->taxonomy );
+		if ( ! $taxonomy ) {
+			return '';
```
It's more conventional to use `null` or `false`.

noisysocks replied:
Done in 0e572f921e.

noisysocks commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,163 @@
+		$namespace = 'wp/v2';
+		$rest_base = ! empty( $taxonomy->rest_base ) ? $taxonomy->rest_base : $taxonomy->name;
+
+		return sprintf( '%s/%s/%d', $namespace, $rest_base, $term->term_id );
```
Can hardcode `wp/v2` instead of using `%s`.

noisysocks replied:
Done in 0e572f921e.

noisysocks committed:
REST API: Add pagination to term search handler

noisysocks committed:
Navigation: Remove unnecessary async

noisysocks committed:
REST API: Fix doc comments

noisysocks committed:
REST API: Use null instead of empty string

noisysocks committed:
REST API: Add pagination to page format search handler

noisysocks committed:
REST API: Fix term search handler tests

noisysocks committed:
REST API: Add post format search handler tests

noisysocks commented:
I've addressed all the feedback here and added unit tests.

@TimothyBJacobs: Would you mind giving this another look?

TimothyBJacobs commented on lib/rest-api.php:
```diff
@@ -330,3 +330,29 @@ function gutenberg_register_image_editor() {
+ * @return array Title updated list of handlers.
+ */
+function gutenberg_post_format_search_handler( $search_handlers ) {
+	$search_handlers[] = new WP_REST_Post_Format_Search_Handler();
```
I think we should probably only register this if `current_theme_supports( 'post-formats' )`.

noisysocks replied:
This means that calling `GET /v2/search` with `type=post-format` will result in a 400 error unless the client checks `OPTIONS /v2/search` first. That's a little cumbersome, no?

TimothyBJacobs replied:
I'm not sure. I think it's probably worth doing an `OPTIONS` request always to be able query other sources that the client doesn't already know about.

But it seems like the preflight check could be skipped since a `400` response with an error for that parameter could be treated as an empty result sense and potentially excluded from future queries.

noisysocks replied:
In d44950bbd3 I made it so that the handler isn't registered and Gutenberg won't call `/v2/search?type=post-format` if the current theme doesn't support post formats.

TimothyBJacobs commented on phpunit/class-wp-rest-post-format-search-handler-test.php:
```diff
@@ -0,0 +1,141 @@
+	 * @param WP_UnitTest_Factory $factory Helper that lets us create fake data.
+	 */
+	public static function wpSetUpBeforeClass( $factory ) {
+		self::$my_post_format_term = $factory->term->create(
```
Can we use `set_post_format` on a created post instead? That feels more durable.

noisysocks replied:
Done in 26e1cb0173.

TimothyBJacobs commented on phpunit/class-wp-rest-post-format-search-handler-test.php:
```diff
@@ -0,0 +1,141 @@
+			)
+		);
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEqualSets(
```
What about checking if the results contains "Aside"? This feels brittle, I think it would fail if there are other post formats created.

noisysocks replied:
Done in 26e1cb0173.

TimothyBJacobs commented on phpunit/class-wp-rest-term-search-handler-test.php:
```diff
@@ -0,0 +1,248 @@
+			wp_list_pluck( $response->get_data(), 'id' )
+		);
+	}
+
```
Can we have a test where the search doesn't match?

noisysocks replied:
Yes we can! 7408a1b861

TimothyBJacobs reviewed (changes requested):
Thanks for making those changes! I've left some more comments.

rickalee commented:
Possible to add author archives while building this?

noisysocks committed:
REST API: Only show term subtypes which are REST API enabled

noisysocks committed:
REST API: Use WP_Term_Query's 'search' param

noisysocks committed:
REST API: Use wp_count_terms() instead of WP_Term_Query

noisysocks committed:
REST API: Set fields=all when fetching terms so that they're cached

noisysocks committed:
REST API: Use rest_get_route_for_term() instead of custom method

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,117 @@
+	 *               total count for the matching search results.
+	 */
+	public function search_items( WP_REST_Request $request ) {
+		$format_strings = get_post_format_strings();
```
I'm not sure. What do you think? Right now it will only return post formats that have posts because of the call to `get_post_format_link()` below.

TimothyBJacobs replied:
I'm not sure either. I think it could go either way. The closest example would be the current nav menu screen which seems to show used formats even if the theme doesn't support them.

noisysocks replied:
Let's try it like this and see if we run into issues. We ostensibly have until 5.6 RC 1 (November 17) to make breaking changes.

noisysocks committed:
REST API: Pass forwards-compatible array to rest_post_format_search_query

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,117 @@
+			$format_link = get_post_format_link( $format_slug );
+			if ( $format_link ) {
+				// Formats don't have an ID, so fake one using the array index.
+				$found_ids[] = $index + 1;
```
It looks like we're only using the ID as a React `key`. So, no, nothing should break if we change the ID to a string.

TimothyBJacobs replied:
Sounds good. I've created a ticket: https://core.trac.wordpress.org/ticket/51131

noisysocks committed:
REST API: Improve stability of WP_REST_Post_Format_Search_Handler tests

noisysocks committed:
REST API: Add test for when no term matches search query

noisysocks commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+		$this->subtypes = array_values(
+			get_taxonomies(
+				array(
+					'public' => true,
```
Good catch. Fixed in 2f596170ed6a71a3f14c71d863e7621117bd9a09.

noisysocks commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+		);
+
+		if ( ! empty( $request['search'] ) ) {
+			$query_args['name__like'] = $request['search'];
```
Changed in fa053dd13d93c199a5efe6ee7fb2e18a281ec835.

noisysocks commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+		unset( $query_args['offset'], $query_args['number'] );
+		$query_args['fields'] = 'count';
+
+		$total = (int) $query->query( $query_args );
```
Done in f6c3618f3e.

noisysocks commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+			'hide_empty' => false,
+			'offset'     => ( $page - 1 ) * $per_page,
+			'number'     => $per_page,
+			'fields'     => 'ids',
```
Good catch! I set `fields` to `'all`' in d09da71b29.

noisysocks commented on lib/class-wp-rest-term-search-handler.php:
```diff
@@ -0,0 +1,162 @@
+
+		$links = array();
+
+		$item_route = $this->detect_rest_item_route( $term );
```
Heck yeah. Done in 4457b4ade3.

noisysocks commented on lib/class-wp-rest-post-format-search-handler.php:
```diff
@@ -0,0 +1,117 @@
+		$format_strings = get_post_format_strings();
+		$format_slugs   = array_keys( $format_strings );
+
+		$query = '';
```
Good plan. Done in fc786cac64.

noisysocks commented:
> Possible to add author archives while building this?

Right now I'm focused on only building out what's necessary for the new Navigation screen in Gutenberg. You're welcome to create a ticket in [Trac](https://core.trac.wordpress.org/) for adding a new author archives search handler. You could copy and paste the code that's in this PR to get a head start on it.

noisysocks commented:
@TimothyBJacobs: Thanks for great feedback! I addressed or responded to everything.

noisysocks committed:
Navigation: Don't call /v2/search?type=post-format if theme does not support formats

TimothyBJacobs reviewed (approved):
REST API wise everything looks good to me!

noisysocks committed:
REST API: Fix post format handler tests
